source/lac/trilinos_tpetra_sparsity_pattern.cc

```diff
@@ -0,0 +1,1078 @@
+            local_entries_per_row_host(i) =
+              n_entries_per_row[row_map->getMinGlobalIndex() + i];
+            total_size += local_entries_per_row_host[i];
+          }
+        local_entries_per_row.modify<Kokkos::DefaultHostExecutionSpace>();
+        local_entries_per_row.sync<Kokkos::DefaultExecutionSpace>();
+
+        AssertThrow(
+          total_size < static_cast<std::uint64_t>(
+                         std::numeric_limits<
+                           dealii::types::signed_global_dof_index>::max()),
+          ExcMessage(
+            "You are requesting to store more elements than global ordinal type allows."));
+
+        graph = Utilities::Trilinos::internal::make_rcp<GraphType<Node>>(
+          row_map, col_map, local_entries_per_row);
+      }
+
+
+
+      template <
+        typename SparsityPatternType,
+        typename Node = Tpetra::KokkosClassic::DefaultNode::DefaultNodeType>
+      void
+      reinit_sp(const Teuchos::RCP<MapType<Node>> &row_map,
+                const Teuchos::RCP<MapType<Node>> &col_map,
+                const SparsityPatternType         &sp,
+                const bool                         exchange_data,
+                Teuchos::RCP<MapType<Node>>       &column_space_map,
+                Teuchos::RCP<GraphType<Node>>     &graph,
+                Teuchos::RCP<GraphType<Node>>     &nonlocal_graph)
+      {
+        nonlocal_graph.reset();
+        graph.reset();
+
+        AssertDimension(sp.n_rows(), row_map->getGlobalNumElements());
+        AssertDimension(sp.n_cols(), col_map->getGlobalNumElements());
+
+        column_space_map =
+          Utilities::Trilinos::internal::make_rcp<MapType<Node>>(*col_map);
+
+        Assert(row_map->isContiguous() == true,
+               ExcMessage(
+                 "This function only works if the row map is contiguous."));
+
+        const size_type<Node> first_row = row_map->getMinGlobalIndex(),
+                              last_row  = row_map->getMaxGlobalIndex() + 1;
+        Teuchos::Array<size_t> n_entries_per_row(last_row - first_row);
+
+        // Trilinos wants the row length as an int this is hopefully never going
+        // to be a problem.
+        for (size_type<Node> row = first_row; row < last_row; ++row)
+          n_entries_per_row[row - first_row] = sp.row_length(row);
+
+        AssertThrow(
+          std::accumulate(n_entries_per_row.begin(),
+                          n_entries_per_row.end(),
+                          std::uint64_t(0)) <
+            static_cast<std::uint64_t>(std::numeric_limits<int>::max()),
+          ExcMessage("The TrilinosWrappers use Tpetra internally which "
+                     "uses 'signed int' to represent local indices. "
+                     "Therefore, only 2,147,483,647 nonzero matrix "
+                     "entries can be stored on a single process, "
+                     "but you are requesting more than that. "
+                     "If possible, use more MPI processes."));
```

Comment:
Not really like this.

It does depend on the template parameters chosen when building Trilinos Tpetra. 
For `local ordinate = int` it still applies but we should think about how to change this assertion
to properly test this.

Reply:
Yeah, we should at least leave a note about recompiling `Trilinos` if users are running into this with `LocalOrdinate=int`.

Reply:
And we should have a guard at the top that makes sure this check is only run if, indeed, we are using 32-bit integers.

Reply:
What do you think about the error message:

```
The TrilinosWrappers use Tpetra internally, and Trilinos/Tpetra was 
compiled with 'local ordinate = int'. Therefore, 'signed int' is used to 
represent local indices, and only 2,147,483,647 nonzero matrix entries 
can be stored on a single process, but you are requesting more than
that. Either use more MPI processes or recompile Trilinos with another 
local ordinate.
```

Reply:
Sounds good to me.

Reply:
Since there are really only two viable options for the local ordinate (iirc), 
we could be a little more specific and say with `local ordnate = long long`
or even state the necessary CMake flag(s) i.e. to disable `LO=int` and enable `LO=long long`
but that would make the message a bit longer.